Plant species mutation fix.

drteaspoon420 opened the pull request:
<!-- Please read these guidelines before opening your PR: https://docs.spacestation14.io/en/getting-started/pr-guideline -->
<!-- The text between the arrows are comments - they will not be visible on your PR. -->

## About the PR
<!-- What did you change in this PR? -->
Species mutation no-longer removes all other mutations.

## Why / Balance
<!-- Why was it changed? Link any discussions or issues here. Please discuss how this would affect game balance. -->
The 'meta' game of botany for anyone who knew how species mutation works would be to dump as much mutagen as you want and use clippers immidietly when species mutated to remove all unwanted mutations such as `Viable`. This also reseted any chemical mutations making it inadvicable to work with any plant that has viable species mutations.

## Technical details
<!-- If this is a code change, summarize at high level how your new code works. This makes it easier to review. -->
`SeedChemQuantity` has new field `Inherent` that is by default true (so no need to update any yml files). This is to keep track of what chemicals to remove when plant experiences 'species' mutation. Any chemical that comes from crossbreeding or mutation has this field set to false (so that the species mutation will not override it).
`SeedData` has new method `SpeciesChange` that take new species prototype `SeedData` and will copy `Name`, `Noun`, `DisplayName`, `Mysterious`, `PacketPrototype`, `ProductPrototypes`, `GrowthStages`, `PlantRsi`, `PlantIconState`, `SplatPrototype` from specified seed, and take rest for original seed. Then goes through chemicals and remove all original chemicals that have `Inherent` as `true` and add all new chemicals from the new species.

## Media
<!-- 
PRs which make ingame changes (adding clothing, items, new features, etc) are required to have media attached that showcase the changes.
Small fixes/refactors are exempt.
Any media may be used in SS14 progress reports, with clear credit given.

If you're unsure whether your PR will require media, ask a maintainer.

Check the box below to confirm that you have in fact seen this (put an X in the brackets, like [X]):
-->

- [X ] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes
<!--
List any breaking changes, including namespace, public class/method/field changes, prototype renames; and provide instructions for fixing them. This will be pasted in #codebase-changes.
-->
None

**Changelog**
<!--
Make players aware of new features and changes that could affect how they play the game by adding a Changelog entry. Please read the Changelog guidelines located at: https://docs.spacestation14.io/en/getting-started/pr-guideline#changelog
-->

:cl: DrTeaSpoon
- fix: Plant species mutation no longer resets other mutations.

Doru991 commented:
To be honest ambrosia deus spam meta was getting a bit boring, I think this is fair enough since it makes trying to get species mutations considerably more tedious (and therefore helps to justify the very good rewards you can get from it).

Doru991 commented:
While I'm not entirely sure about this, I would maybe suggest bumping species mutation's probability up to 10 'bits' from 5, in order to give a slightly higher chance that the plant won't be totally messed up by the time a species mutation is triggered.

Partmedia commented on Content.Server/Botany/SeedPrototype.cs:
```diff
@@ -59,6 +59,8 @@ public partial struct SeedChemQuantity
     [DataField("Max")] public int Max;
     [DataField("PotencyDivisor")] public int PotencyDivisor;
+    [DataField("Inherent")] public bool Inherent = true;
+
```
Extra space

drteaspoon420 replied:
fixed

Partmedia commented on Content.Server/Botany/SeedPrototype.cs:
```diff
@@ -59,6 +59,8 @@ public partial struct SeedChemQuantity
     [DataField("Min")] public int Min;
     [DataField("Max")] public int Max;
     [DataField("PotencyDivisor")] public int PotencyDivisor;
+    [DataField("Inherent")] public bool Inherent = true;
```
Please start documenting new fields with the explanations you wrote in "Technical Details". It should be easier for someone reading the source in the future to understand what these new fields mean/do.

drteaspoon420 replied:
fixed

Partmedia commented on Content.Server/Botany/SeedPrototype.cs:
```diff
@@ -293,4 +295,76 @@ public SeedData Clone()
         return newSeed;
     }
+
+    public SeedData SpeciesChange(SeedData other)
```
Missing method documentation

drteaspoon420 replied:
Fixed

Partmedia commented on Content.Server/Botany/Systems/MutationSystem.cs:
```diff
@@ -295,10 +297,8 @@ private void MutateSpecies(ref SeedData seed, int bits, int totalbits, float mul
-        seed.Potency = oldSeed.Potency;
-        seed.Yield = oldSeed.Yield;
+        seed = seed.SpeciesChange(protoSeed);
+
```
Extra space

Partmedia commented on Content.Server/Botany/SeedPrototype.cs:
```diff
@@ -293,4 +295,76 @@ public SeedData Clone()
+            Chemicals.TryAdd(otherChem.Key, otherChem.Value);
+        }
+
+        foreach (var originalChem in Chemicals)
```
Add comment explaining what's going on

Partmedia reviewed (changes requested):
Thank you for your contribution. I have a few nitpicks that I'd like you to take a look at.

Partmedia commented:
:cl: "Plant species mutation no-longer resets other mutations."
"no-longer" is not hyphenated, it should be "no longer".

drteaspoon420 commented:
I have addressed the issues. Code has now proper commentary. Additionally commented the other seed chemical things because they would slightly confusing, if you didn't go and read the code where things like `PotencyDivisor` is used.

Partmedia commented:
In the future, please include screenshots or videos to demonstrate testing done.

Aisu9 commented:
this change made impossible to get desired chemical from mutation. ambrosia deus for omnizine and koibean for carpotoxin. making theses unique mutation plants useless in what they are suppose to provide

Partmedia commented:
> this change made impossible to get desired chemical from mutation. ambrosia deus for omnizine and koibean for carpotoxin. making theses unique mutation plants useless in what they are suppose to provide

Thank you for the report. @drteaspoon420 would you be able to take a look?

Doru991 commented on Content.Server/Botany/SeedPrototype.cs:
```diff
@@ -293,4 +310,82 @@ public SeedData Clone()
+        }
+
+        // Removing the inherent chemicals from the old species. Leaving mutated/crossbread ones intact.
+        foreach (var originalChem in Chemicals)
+        {
+            if (!other.Chemicals.ContainsKey(originalChem.Key) && originalChem.Value.Inherent)
+            {
+                Chemicals.Remove(originalChem.Key);
+            }
+        }
```
I think this is where the issue could be, since we remove inherent chems AFTER adding them, so the inherent chems of the new might species also get cut

Doru991 replied:
Disregard that, I misread something

Doru991 replied:
I think I found it for real this time
